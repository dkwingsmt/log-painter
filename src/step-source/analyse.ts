import {
  AnalysedPlayer,
  GroupResult,
  AnalysedLine,
} from 'step-source';
// import {
//   bbsColors,
//   DescribedColor,
// } from './colors';
import {
  ConfigPlayer,
} from 'step-config';

export interface AnalyseResult {
  lines: AnalysedLine[];
  playerIds: string[];
}

// const presetColors: string[] = bbsColors.map(({ value }: DescribedColor) => value);

function findFirstPlayerConfig(
  configPlayers: Record<string, ConfigPlayer>,
  playerIdGroup: string[],
): ConfigPlayer | null {
  for (const playerId of playerIdGroup) {
    if (configPlayers[playerId])
      return configPlayers[playerId];
  }
  return null;
}

export function analyse(grouped: GroupResult): AnalyseResult {
  const { players, lines } = grouped;
  const configPlayers: Record<string, ConfigPlayer> = {};
  const currentPlayers: Record<string, ConfigPlayer> = {};
  for (const player of players) {
    let configPlayer = findFirstPlayerConfig(configPlayers, player.allPlayerIds);
    if (!configPlayer) {
      configPlayer = {
        enabled: true,
        displayName: player.name,
        color: 'black',
      };
    }
    for (const playerId of player.allPlayerIds) {
      currentPlayers[playerId] = configPlayer;
    }
  };
  return {
    lines,
    playerIds: players.map((player: AnalysedPlayer) => player.playerId),
  };
}