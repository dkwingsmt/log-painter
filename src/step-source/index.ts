export type {
  AnalysedPlayer,
  AnalysedLine,
  GroupResult,
} from './group';

export {
  StepSource,
} from './StepSource';
export type {
  StepSourceResult,
} from './StepSource';