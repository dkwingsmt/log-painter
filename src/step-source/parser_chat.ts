import compact from 'lodash/compact';
import trim from 'lodash/trim';
import flow from 'lodash/flow';

export interface ParsedPlayer {
  name: string;
  number?: string;
  title?: string;
}

export interface ParsedLine {
  time?: string;
  player: ParsedPlayer;
  content: string[];
}

export interface ParseResult {
  logLines: ParsedLine[];
}

interface ParsedHeader {
  player: ParsedPlayer;
  time?: string;
  content?: string;
}

type HeaderParser = (line: string) => ParsedHeader | null;
type LogLineConverter = (logLine: ParsedLine) => ParsedLine | null;

interface LogConfig {
  headerParser: HeaderParser;
  logLineConverter: LogLineConverter;
}

function removeSystemTextConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;

  const withdrawParser = /^.*撤回了一条消息/;
  const withdrawEnglishParser = /^.* recalled a message( Re-edit)?$/;
  const withdrawMemberParser = /^.*撤回了成员.*的一条消息$/;
  const withdrawMember2Parser = /^.*撤回了一条成员消息$/;
  const commonFriendsParser = /^你和.*有\d+个共同好友，点击添加好友。$/;
  const joinMemberParser = /^.*加入本群。$/;
  const inviteMemberParser = /^.*邀请.*加入了本群。$/;

  let end = lines.length - 1;
  for (; end >=0; end--) {
    if ([
      /^ *$/,
      withdrawParser,
      withdrawEnglishParser,
      withdrawMemberParser,
      withdrawMember2Parser,
      commonFriendsParser,
      joinMemberParser,
      inviteMemberParser,
    ].some((toMatch: RegExp) => toMatch.exec(lines[end]))) {
      continue;
    }
    break;
  }
  logLine.content = logLine.content.slice(0, end + 1);
  return logLine;
}

const regNumber = /\(\d+\)|<.+@.+\..+>/;
const regTitle = /(?:【(.{1,6})】)?/;
const regTime = /(?:(?:上午|下午) )?\d{1,2}:\d{2}:\d{2}(?: (?:AM|PM))?/;

function removeMobileDateConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;

  const dateParser = /^— —  \d{4}-\d{1,2}-\d{1,2}  — —$/;

  let end = lines.length - 1;
  for (; end >=0; end--) {
    if ([
      /^ *$/,
      dateParser,
    ].some((toMatch: RegExp) => toMatch.exec(lines[end]))) {
      continue;
    }
    break;
  }
  logLine.content = logLine.content.slice(0, end + 1);
  return logLine;
}

function removeMessageManagerSystemTextConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;

  const dateParser = /^ \d{4}-\d{2}-\d{2}$/;
  const withdrawParser = new RegExp(`^${regTime.source}.*撤回了一条消息`);
  const withdrawEnglishParser = new RegExp(`^${regTime.source}.* recalled a message( Re-edit)?`);
  const withdrawMemberParser = new RegExp(`^${regTime.source}.*撤回了成员.*的一条消息$`);
  const withdrawMember2Parser = new RegExp(`^${regTime.source}.*撤回了一条成员消息$`);
  // const commonFriendsParser = /^你和.*有\d+个共同好友，点击添加好友。$/;
  const joinMemberParser = new RegExp(`^${regTime.source}.*加入本群。$`);
  // const inviteMemberParser = /^.*邀请.*加入了本群。$/;

  let end = lines.length - 1;
  for (; end >=0; end--) {
    if ([
      /^ *$/,
      dateParser,
      withdrawParser,
      withdrawEnglishParser,
      withdrawMemberParser,
      withdrawMember2Parser,
      joinMemberParser,
    ].some((toMatch: RegExp) => toMatch.exec(lines[end]))) {
      continue;
    }
    break;
  }
  logLine.content = logLine.content.slice(0, end + 1);
  return logLine;
}

function defaultConverter(logLine: ParsedLine | null): ParsedLine | null {
  if (logLine == null)
    return null;
  const lines = logLine.content;
  // Remove starting and trailing empty lines.
  while(lines.length > 0 && lines[0].trim() === '') {
    lines.splice(0, 1);
  }
  while(lines.length > 0 && lines[lines.length-1].trim() === '') {
    lines.splice(lines.length-1, 1);
  }
  if (lines.length === 0)
    return null;
  return logLine;
}

function trimNumber(src: string): string {
  return trim(src, '()<>');
}

// Export from log
// E.g. "2019-09-23 8:43:38 PM 骰娘-Roll100(872001750)"
// E.g. "2019-09-23 8:43:38 PM 骰娘-Roll100"
const exportFromLog: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regDateName = new RegExp(`\\d{4}-\\d{2}-\\d{2} ${regTime.source}`);
    const regHeader = new RegExp(`^(${regDateName.source}) (.*?)(${regNumber.source})?$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, time, name, number] = matches;
    return {
      player: {
        name,
        number: number == null ? undefined : trimNumber(number),
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    if (['10000', '1000000'].includes(logLine.player.number || ''))
      return null;
    return defaultConverter(logLine);
  },
};

// Copy from chat mainwindow or sidewindow
// E.g. "【冒泡】无情的围观熊 2/14/2020 9:16:13 PM"
// E.g. "【煤油】丧 丧 熊 9:59:54 PM"
const copyFromChat: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regDateTime = new RegExp(`(?:\\d{1,4}\\/\\d{1,2}\\/\\d{1,4} (?:星期. )?)?${regTime.source}`);
    const regHeader = new RegExp(`^${regTitle.source}(.*?) (${regDateTime.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, title, name, time] = matches;
    return {
      player: {
        name,
        title,
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeSystemTextConverter,
      defaultConverter,
    )(logLine);
  },
};

// Copy from message manager
// E.g. "织练取(958884) 3:13:28 AM"
const copyFromMessageManager: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = new RegExp(`^${regTitle.source}(.*?)(${regNumber.source}) (${regTime.source})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, title, name, number, time] = matches;
    return {
      player: {
        name,
        title,
        number: trimNumber(number),
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeMessageManagerSystemTextConverter,
      defaultConverter,
    )(logLine);
  },
};

// Copy from mobile
// E.g. "a dark ideation  23:50:40"
const copyFromMobile: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = /^(.*?) {2}(\d{1,2}:\d{2}:\d{2}) *$/;
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, name, time] = matches;
    return {
      player: {
        name,
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      removeMobileDateConverter,
      defaultConverter,
    )(logLine);
  },
};

// 溯洄骰的Log记录器
// E.g. "名字(123456) 2021-11-08 04:07:37"
const shikiLogRecord: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = new RegExp(`^(.*?)(${regNumber.source}) (\\d{4}-\\d{2}-\\d{2} \\d{2}:\\d{2}:\\d{2})$`);
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, name, number, time] = matches;
    return {
      player: {
        name,
        number: trimNumber(number),
      },
      time,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      defaultConverter,
    )(logLine);
  },
};

// Reparse log with angular brackes
// E.g. "<小明> 内容内容内容内容"
const reparseAngularBracket: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = /^<(.+)> (.+)$/;
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, name, content] = matches;
    return {
      player: {
        name,
      },
      content,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      defaultConverter,
    )(logLine);
  },
};

// Reparse log with bold brackes
// E.g. "【小明】内容内容内容内容"
const reparseBoldBracket: LogConfig = {
  headerParser: (line: string): ParsedHeader | null => {
    const regHeader = /^【(.+)】(.+)$/;
    const matches = regHeader.exec(line);
    if (!matches)
      return null;
    const [_all, name, content] = matches;
    return {
      player: {
        name,
      },
      content,
    };
  },
  logLineConverter: (logLine: ParsedLine): ParsedLine | null => {
    return flow(
      defaultConverter,
    )(logLine);
  },
};

export function parseChat(data: string): ParseResult {
  const logLines: ParsedLine[] = [];
  let firstLogConfig: LogConfig | undefined = undefined;
  for (const line of data.split('\n')) {
    const parsedHeader = ((): ParsedHeader | null => {
      if (firstLogConfig)
        return firstLogConfig.headerParser(line);
      for (const logConfig of [
        shikiLogRecord,
        copyFromMessageManager,
        copyFromMobile,
        exportFromLog,
        copyFromChat,
        reparseAngularBracket,
        reparseBoldBracket,
      ]) {
        const result = logConfig.headerParser(line);
        if (result) {
          firstLogConfig = logConfig;
          return result;
        }
      }
      return null;
    })();
    if (parsedHeader) {
      const { player, time, content } = parsedHeader;
      logLines.push({
        player,
        time,
        content: content ? [content] : [],
      });
    } else {
      if (firstLogConfig) {
        const currentLogLine: ParsedLine | undefined = logLines[logLines.length - 1];
        // Content after a recognizable header
        if (currentLogLine)
          currentLogLine.content.push(line);
      } else {
        // Unrecognizable content before the first header
        // Do nothing
      }
    }
  }
  if (!firstLogConfig)
    return { logLines: [] };

  const convertedLogLines = compact(
    logLines.map(firstLogConfig.logLineConverter)
  );
  return { logLines: convertedLogLines };
}
