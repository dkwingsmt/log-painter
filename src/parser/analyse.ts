import forEach from 'lodash/forEach';
import values from 'lodash/values';
import clone from 'lodash/clone';
import { ParseResult, ParsedPlayer } from './parser_chat';
import {
  Configuration,
  AnalysedLine,
  AnalysedPlayer,
  AnalyseResult,
  ConfigPlayer,
  presetDescribedColors,
  DescribedColor,
} from 'common';

interface GroupResult {
  players: AnalysedPlayer[];
  lines: AnalysedLine[];
};

const presetColors: string[] = presetDescribedColors.map(({ value }: DescribedColor) => value);

function getPlayerIdGroup(player: ParsedPlayer): string[] {
  const ids: string[] = [];
  if (player.number)
    ids.push(`qq:${player.number}`);
  ids.push(`name:${player.name}`);
  return ids;
}

function groupByPlayers(result: ParseResult): GroupResult {
  const lines: AnalysedLine[] = [];
  const players: Record<string, AnalysedPlayer> = {};
  for (const parsedLine of result.logLines) {
    const { player, time, content } = parsedLine;
    const playerIdGroup = getPlayerIdGroup(player);
    const majorPlayerId = playerIdGroup[0];
    if (!players[majorPlayerId])
      players[majorPlayerId] = {
        playerId: majorPlayerId,
        allPlayerIds: playerIdGroup,
        name: player.name,
        number: player.number,
      };
    lines.push({
      playerId: majorPlayerId,
      time,
      title: player.title,
      content,
    });
  }
  return {
    lines,
    players: values(players),
  };
}

function findFirstPlayerConfig(
  configPlayers: Record<string, ConfigPlayer>,
  playerIdGroup: string[],
): ConfigPlayer | null {
  for (const playerId of playerIdGroup) {
    if (configPlayers[playerId])
      return configPlayers[playerId];
  }
  return null;
}

function findFirstUnusedPresetColor(
  presetColors: string[],
  configPlayers: Record<string, ConfigPlayer>,
): string {
  const existingColors: Record<string, boolean> = {};
  forEach(configPlayers, (configPlayer: ConfigPlayer) => {
    existingColors[configPlayer.color] = true;
  });
  return presetColors.find((color: string) => !existingColors[color]) || 'black';
}

export function analyse(parsed: ParseResult, currentConfig: Configuration): AnalyseResult {
  const { lines, players } = groupByPlayers(parsed);
  const configPlayers: Record<string, ConfigPlayer> = clone(currentConfig.players);
  for (const player of players) {
    let configPlayer = findFirstPlayerConfig(configPlayers, player.allPlayerIds);
    if (!configPlayer) {
      configPlayer = {
        enabled: true,
        displayName: player.name,
        color: findFirstUnusedPresetColor(presetColors, configPlayers),
      };
    }
    for (const playerId of player.allPlayerIds) {
      configPlayers[playerId] = configPlayer;
    }
  };
  return {
    lines,
    playerIds: players.map((player: AnalysedPlayer) => player.playerId),
    nextConfig: {
      players: configPlayers,
    },
  };
}