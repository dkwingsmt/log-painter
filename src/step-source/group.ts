import values from 'lodash/values';

import { parseChat, ParseResult, ParsedPlayer } from './parser_chat';

export interface AnalysedLine {
  playerId: string;
  time?: string;
  title?: string;
  content: string[];
}

export interface AnalysedPlayer {
  playerId: string;
  allPlayerIds: string[];
  number?: string;
  name: string;
}

export interface GroupResult {
  players: AnalysedPlayer[];
  lines: AnalysedLine[];
};

function getPlayerIdGroup(player: ParsedPlayer): string[] {
  const ids: string[] = [];
  if (player.number)
    ids.push(`qq:${player.number}`);
  ids.push(`name:${player.name}`);
  return ids;
}

function groupByPlayers(result: ParseResult): GroupResult {
  const lines: AnalysedLine[] = [];
  const players: Record<string, AnalysedPlayer> = {};
  for (const parsedLine of result.logLines) {
    const { player, time, content } = parsedLine;
    const playerIdGroup = getPlayerIdGroup(player);
    const majorPlayerId = playerIdGroup[0];
    if (!players[majorPlayerId])
      players[majorPlayerId] = {
        playerId: majorPlayerId,
        allPlayerIds: playerIdGroup,
        name: player.name,
        number: player.number,
      };
    lines.push({
      playerId: majorPlayerId,
      time,
      title: player.title,
      content,
    });
  }
  return {
    lines,
    players: values(players),
  };
}

export function parseAndGroup(source: string): GroupResult {
  const parseResult: ParseResult = parseChat(source);
  return groupByPlayers(parseResult);
}