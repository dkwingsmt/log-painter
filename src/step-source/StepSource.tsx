import React, { useState, useContext } from 'react';

import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';

import {
  useStepperStyles,
  AlertDialog,
  AlertDialogControl,
  StartStepProps,
  Configuration,
  configContext,
} from 'common';
import { parseAndGroup } from './group';
import { analyse, AnalyseResult } from './analyse';

export type StepSourceResult = AnalyseResult;

type StepSourceProps = StartStepProps<StepSourceResult, Configuration>;


function processResult(text: string): AnalyseResult | undefined {
  const groupResult = parseAndGroup(text);
  if (groupResult.players.length === 0 || groupResult.lines.length === 0) {
    return undefined;
  }
  return analyse(groupResult);
}

export const StepSource: React.FC<StepSourceProps> = (props: StepSourceProps) => {
  const { onNextStep } = props;
  const config: Configuration = useContext<Configuration>(configContext);
  const stepperClasses = useStepperStyles();
  const [text, setText] = useState<string>('');

  const [alertControl, setAlertControl] = useState<AlertDialogControl>({
    open: false,
  });

  if (!props.show)
    return null;

  return (
    <Grid container className={stepperClasses.Container}>
      <Grid item xs={12} className={stepperClasses.Body}>
        <TextField
          id="outlined-basic"
          label="原文"
          placeholder="把原文粘贴到这里，然后点击下一步。"
          variant="outlined"
          multiline
          fullWidth
          rows={10}
          className='Source-textfield'
          onChange={(event: React.ChangeEvent<HTMLInputElement>): void => {
            setText(event.target.value);
          }}
          value={text}
        />
      </Grid>
      <Grid container item xs={12} justifyContent="flex-end" className={stepperClasses.Control}>
        <Button
          variant="contained"
          color="primary"
          disabled={text.length < 5}
          className={stepperClasses.ControlButton}
          onClick={(): void => {
            const result = processResult(text);
            if (result == null) {
              setAlertControl({
                open: true,
                body: '无法从这段记录中找到可识别的部分！请检查后再试一次吧！',
              });
            } else {
              onNextStep(result, config);
            }
          }}
        >
          下一步
        </Button>
      </Grid>
      <AlertDialog
        onClose={(): void => setAlertControl({
          ...alertControl,
          open: false,
        })}
        {...alertControl}
      />
    </Grid>
  );
};
