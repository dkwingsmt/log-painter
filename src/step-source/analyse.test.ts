import { GroupResult, AnalysedPlayer } from 'step-source';
import { analyse, AnalyseResult } from './analyse';

function setOf(elements: string[]): Set<string> {
  const result = new Set<string>();
  for (const element of elements) {
    result.add(element);
  }
  return result;
}

it('should work for empty config', () => {
  const player1: AnalysedPlayer = {
    playerId: 'name:a',
    names: setOf(['a']),
  };
  const player2: AnalysedPlayer = {
    playerId: 'name:b',
    names: setOf(['b']),
  };
  const groupResult: GroupResult = {
    players: [
      player1,
      player2,
    ],
    lines: [
      {
        time: 'time1',
        playerId: player1.playerId,
        content: ['1', '2'],
        title: 'A',
      },
      {
        time: 'time2',
        playerId: player2.playerId,
        content: ['3'],
        title: 'B',
      },
      {
        time: 'time3',
        playerId: player1.playerId,
        content: ['4'],
        title: 'A',
      },
    ]
  };
  const analyseResult: AnalyseResult = analyse(groupResult);
  expect(analyseResult).toMatchSnapshot();
});
