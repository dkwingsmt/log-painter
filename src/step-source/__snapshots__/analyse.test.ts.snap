// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should work for empty config 1`] = `
Object {
  "lines": Array [
    Object {
      "content": Array [
        "1",
        "2",
      ],
      "playerId": "name:a",
      "time": "time1",
      "title": "A",
    },
    Object {
      "content": Array [
        "3",
      ],
      "playerId": "name:b",
      "time": "time2",
      "title": "B",
    },
    Object {
      "content": Array [
        "4",
      ],
      "playerId": "name:a",
      "time": "time3",
      "title": "A",
    },
  ],
  "playerIds": Array [
    "name:a",
    "name:b",
  ],
}
`;

exports[`should work when some players have filled starting colors 1`] = `
Object {
  "lines": Array [
    Object {
      "content": Array [
        "1",
        "2",
      ],
      "playerId": "name:a",
      "time": "time1",
      "title": "A",
    },
    Object {
      "content": Array [
        "3",
      ],
      "playerId": "name:b",
      "time": "time2",
      "title": "B",
    },
  ],
  "playerIds": Array [
    "name:a",
    "name:b",
  ],
}
`;
